gapil/runtime/cc/map.inc

```diff
@@ -25,6 +26,99 @@
 static_assert((GAPIL_MIN_MAP_SIZE & (GAPIL_MIN_MAP_SIZE - 1)) == 0, "Map size must be a power of 2");
 static_assert((GAPIL_MAP_GROW_MULTIPLIER & (GAPIL_MAP_GROW_MULTIPLIER - 1)) == 0, "Map size must be a power of 2");
 
+// Note these implementations should match map.go
+
+namespace {
+struct fixedValue {
+    uint64_t fixed64;
+};
+
+template<typename T>
+struct gapil_hash {
+    uint64_t operator()(const T& t) {
+        GAPID_ASSERT_MSG(false, "unknown hash type");
+        return 0;
+    }
+};
+
+uint64_t rotate_right(const uint64_t& v, const uint64_t bits) {
+    return (v >> bits) | (v << (64-bits));
+}
+
+uint64_t shift_left(const uint64_t& v, const uint64_t bits) {
+    return v << bits;
+}
+
+template<>
+struct gapil_hash<fixedValue> {
+    uint64_t operator() (const fixedValue& val) {
+        uint64_t v = val.fixed64;
+        v += shift_left(v, 21);
+        v ^= rotate_right(v, 24);
+        v += shift_left(v, 3) + shift_left(v, 8);
+        v ^= rotate_right(v, 14);
+        v += shift_left(v, 2) + shift_left(v, 4);
+        v ^= rotate_right(v, 28);
+        v += shift_left(v, 31);
+        return v;
+    }
+};
+
+template<>
+struct gapil_hash<float> {
+    uint64_t operator() (const float& f) {
+        return gapil_hash<fixedValue>()(fixedValue{static_cast<uint64_t>(*reinterpret_cast<const uint32_t*>(&f))});
```

Comment:
I believe for this sort of thing you're supposed to use a union.

Reply:
I think actually both end up being wrong for the same reason. (memory alignment :\) if alignof(float) != alignof(uint32_t).

I will do the "right thing" and go through a memcpy, which as far as I can tell is the only safe way to do this in general.